test/deploy.py

```diff
@@ -0,0 +1,569 @@
+        """Create a virtual machine for LXD."""
+        if not self.client:
+            raise AssertionError("LXD Client not available to runner.")
+
+        # Create Virtual Machine
+        vm_name = self.deploy_tag + "-" + _get_random_string(4)
+        config = {
+            "name": vm_name,
+            "instance_type": flavor,
+            "storage": pool_name,
+            "profiles": [profile_name],
+            "devices": {
+                "root": {
+                    "path": "/",
+                    "pool": pool_name,
+                    "size": "20GB",
+                    "type": "disk"
+                }
+            },
+            "source": {
+                "type": "image",
+                "certificate": "",
+                "alias": image,
+                "server": "https://images.linuxcontainers.org",
+                "protocol": "simplestreams",
+                "mode": "pull",
+                "allow_inconsistent": False,
+            },
+        }
+
+        # Create VM
+        print("Creating VM {}".format(vm_name))
+        self.client.virtual_machines.create(config, wait=True)
+        self.model["vm_name"] = vm_name
+
+        if is_start:
+            self.client.virtual_machines.get(vm_name).start(wait=True)
+
+        self.wait_for_vm_ready(vm_name)
+        # Increase Root Partition size on VM.
+        self.grow_root_partition(vm_name)
+        return vm_name  # vm_name to refer to the newly create vm.
+
+    def check_call_on_vm(
+        self, vm_name: string, cmd: list, is_fail_print=True
+    ) -> tuple:
+        """Execute Command on VM."""
+        if not self.client:
+            raise AssertionError("LXD Client not available to runner.")
+
+        if not self.client.instances.exists(vm_name):
+            raise AssertionError("VM {} does not exist.".format(vm_name))
+
+        inner_cmd = ["lxc", "exec", vm_name, "--", *cmd]
+        try:
+            subprocess.check_call(inner_cmd)
+        except subprocess.CalledProcessError as e:
+            if is_fail_print:
+                print("Failed Executing on {}: Output {}".format(vm_name, e))
+            raise e
+
+    def check_output_on_cephadm_shell(
+        self, vm_name: string, cmd: list, is_fail_print=True
+    ) -> dict:
+        """Execute cmd on cephadm and return output"""
+        if not self.client:
+            raise AssertionError("LXD Client not available to runner.")
+
+        if not self.client.instances.exists(vm_name):
+            raise AssertionError("VM {} does not exist.".format(vm_name))
+
+        inner_cmd = ["lxc", "exec", vm_name, "--", "cephadm", "shell", *cmd]
+        try:
+            return subprocess.check_output(inner_cmd).decode("UTF-8")
+        except subprocess.CalledProcessError as e:
+            if is_fail_print:
+                print("Failed Cephadm Execution on {}: Output {}"
+                      .format(vm_name, e))
+            raise e
+
+    def wait_for_vm_ready(self, vm_name) -> None:
```

Comment:
Should this function wait forever ? I think bailing out after some point makes sense.

Reply:
Yeah, in case the host is oversubscribed, the VM may never come up. I will add a 10 attempt bailing point.